fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonMetadataCache.java

```diff
@@ -0,0 +1,219 @@
+
+        CacheFactory schemaCacheFactory = new CacheFactory(
+                OptionalLong.of(28800L),
+                OptionalLong.of(Config.external_cache_expire_time_minutes_after_access * 60),
+                Config.max_external_table_cache_num,
+                true,
+                null);
+        this.schemaCache = schemaCacheFactory.buildCache(key -> loadSchema(key), null, executor);
+    }
+
+    @NotNull
+    private PaimonSchemaCacheValue loadSchema(PaimonSchemaCacheKey key) {
+        try {
+            PaimonSchema schema = loadPaimonSchemaBySchemaId(key);
+            List<DataField> columns = schema.getFields();
+            List<Column> dorisColumns = Lists.newArrayListWithCapacity(columns.size());
+            Set<String> partitionColumnNames = Sets.newHashSet(schema.getPartitionKeys());
+            List<Column> partitionColumns = Lists.newArrayList();
+            for (DataField field : columns) {
+                Column column = new Column(field.name().toLowerCase(),
+                        PaimonUtil.paimonTypeToDorisType(field.type()), true, null, true, field.description(), true,
+                        field.id());
+                dorisColumns.add(column);
+                if (partitionColumnNames.contains(field.name())) {
+                    partitionColumns.add(column);
+                }
+            }
+            return new PaimonSchemaCacheValue(dorisColumns, partitionColumns);
+        } catch (IOException e) {
+            throw new CacheException("failed to loadSchema for: %s.%s.%s.%s",
+                    e, key.getCatalog().getName(), key.getDbName(), key.getTableName(), key.getSchemaId());
+        }
+    }
+
+    private PaimonSchema loadPaimonSchemaBySchemaId(PaimonSchemaCacheKey key) throws IOException {
+        Table table = ((PaimonExternalCatalog) key.getCatalog()).getPaimonTable(key.getDbName(),
+                key.getTableName() + Catalog.SYSTEM_TABLE_SPLITTER + SchemasTable.SCHEMAS);
+        PredicateBuilder builder = new PredicateBuilder(table.rowType());
+        Predicate predicate = builder.equal(0, key.getSchemaId());
+        // Adding predicates will also return excess data
+        List<InternalRow> rows = PaimonUtil.read(table, new int[][] {{0}, {1}, {2}}, predicate);
+        for (InternalRow row : rows) {
+            PaimonSchema schema = PaimonUtil.rowToSchema(row);
+            if (schema.getSchemaId() == key.getSchemaId()) {
+                return schema;
+            }
+        }
+        throw new CacheException("failed to loadSchema for: %s.%s.%s.%s",
+                null, key.getCatalog().getName(), key.getDbName(), key.getTableName(), key.getSchemaId());
+    }
+
+    @NotNull
+    private PaimonSnapshotCacheValue loadSnapshot(PaimonSnapshotCacheKey key) {
+        try {
+            PaimonSnapshot latestSnapshot = loadLatestSnapshot(key);
+            List<Column> partitionColumns = getPaimonSchema(key.getCatalog(), key.getDbName(), key.getTableName(),
```

Comment:
I think this is not safe to get value from another cache when loading cache.
Both cache use some executor pool and may becausing dead lock